src/java.base/share/classes/jdk/internal/foreign/layout/LayoutTransformers.java

```diff
@@ -0,0 +1,261 @@
+ *  under the terms of the GNU General Public License version 2 only, as
+ *  published by the Free Software Foundation.  Oracle designates this
+ *  particular file as subject to the "Classpath" exception as provided
+ *  by Oracle in the LICENSE file that accompanied this code.
+ *
+ *  This code is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ *  version 2 for more details (a copy is included in the LICENSE file that
+ *  accompanied this code).
+ *
+ *  You should have received a copy of the GNU General Public License version
+ *  2 along with this work; if not, write to the Free Software Foundation,
+ *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ *
+ */
+
+package jdk.internal.foreign.layout;
+
+import jdk.internal.ValueBased;
+
+import java.lang.foreign.AddressLayout;
+import java.lang.foreign.GroupLayout;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.PaddingLayout;
+import java.lang.foreign.SequenceLayout;
+import java.lang.foreign.StructLayout;
+import java.lang.foreign.UnionLayout;
+import java.lang.foreign.ValueLayout;
+import java.nio.ByteOrder;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * A JDK-internal utility class that provides useful layout transformations.
+ */
+public final class LayoutTransformers {
+
+    private LayoutTransformers() {}
+
+    /**
+     * {@return a function that recursively will remove all member names in memory layouts}
+     */
+    public static Function<MemoryLayout, MemoryLayout> removeNames() {
+        return LayoutTransformerImpl.REMOVE_NAMES::deepTransform;
+    }
+
+    /**
+     * {@return a transformer that recursively will set byte ordering to the provided
+     * {@code byteOrder}}
+     */
+    public static Function<MemoryLayout, MemoryLayout> setByteOrder(ByteOrder byteOrder) {
+        return LayoutTransformerImpl.setByteOrder(byteOrder)::deepTransform;
+    }
+
+    // Internal classes
+
+    /**
+     * A layout transformer that can be used to apply functions on selected types
+     * of memory layouts.
+     * <p>
+     * A layout transformer can be used to convert byte ordering for all sub-members
+     * or remove all names, for example.
+     *
+     * @param <T> the type of MemoryLayout for which transformation is to be made
+     */
+    sealed interface LayoutTransformer<T extends MemoryLayout>
+            permits LayoutTransformerImpl {
+
+        /**
+         * {@return a transformed version of the provided {@code layout}}.
+         * @param layout to transform
+         */
+        MemoryLayout transform(T layout);
```

Comment:
Not too sure about having a distinction between transform and deepTransform, at least not for the use cases we have so far. I'd say just keep the semantics of deepTransform, so that this becomes a functional interface?